### PR TITLE
Use pandoc instead of make4ht

### PR DESCRIPTION
**Pull request by gnuhead-chieb**
Please don't merge PR for now.This PR is in progress now.
I have to fix more issues and create translation process

### CODE REVIEW

**Comment by gnuhead-chieb** [comment]
Milestones

- [X] Custom color name does not work -Fixed
- [X] Icon doesn't show correctly -Fixed
- [X] Alertbox doesn't show correctly -Fixed
- [X]Label hyperref broken in pages/app-details-page.tex and pages/settings-page.tex -Fixed
- [ ] Listing doesn't working in appendices/app_ops.tex(C.3.3=7.3.3)
- [ ] Add link on section title
- [ ] Hide level5 section number
- [ ] Add doctool.sh commanline controller
- [ ] Add string key for making translatable with xliff ( %%!! string_key<< STRING %%!!>>)
- [ ] Extract string and update xliff translation file
- [ ] Restore TEX from xliff

**Review by Alexisblues — APPROVED** [review]
```

```

**Comment by gnuhead-chieb** [comment]
Does anyone have an idea to merge XML translation(content) file to Tex?
Source Tex Format
```
\(section|subsection|subsubsection){SECTION TITLE} %%##title_key==title>>
%%!!string_key<<
CONTENTCONTENTCONTENT
CONTENTCONTENTCONTENT
%%!!>>
```

XML Format
```
<?xml version="1.0" encoding="utf-8"?>
<resources xmlns:xliff="urn:oasis:names:tc:xliff:document:1.2">
<string name=title_key==title">SECTION TITLE</string>
...
<string name="faq_aot-usbdebugging">CONTENTCONTENTCONTENT
CONTENTCONTENTCONTENT</string>
</resources>
```

Strings extraction script from XML Translation file
```
keys=$(grep -oP "(?<=<string name=\").*?(?=\">)" strings.xml)
while read key_title
do

string_title=$(echo 'cat resources/string[@name="'${key_title}'"]/text()' | xmllint --shell strings.xml | sed -e '$d' -e '1d')

done < <(echo "$keys" | grep -P ".*(?===title)")

while read key_content
do

string_content=$(echo 'cat resources/string[@name="'${key_content}'"]/text()' | xmllint --shell strings.xml | sed -e '$d' -e '1d')

done < <(echo "$keys" | grep -Pv ".*(?===title)")
```

**Comment by gnuhead-chieb** [comment]
@MuntashirAkon 
I've done most of process.Before I start finalization,Please check this PR to work script in your environment or not.

Please note that this script is require **bash**(Not work with sh) and **GNU Sed**(Not posix).
and you have to install or put [pandoc-crossref](https://github.com/lierdakil/pandoc-crossref/releases) binary into working directory(This binary isn't include in PR).Other dependencies is written in "doctool.sh help" option.

You can use strings_sampleja.xml for check translation process.

**Review by MuntashirAkon — COMMENTED** [review]
A few questions for you:
1. Can we still build a pdf file? (It's not important, just for understanding)
2. Does every conversion process preserve the file structure i.e. is the output file identical upon subsequent build? This is very important otherwise we wouldn't be able reproduce the same binary file with the same set of configuration.
3. What's the state of the pdf files? Are the now converted into PNG or something?

PS: I haven't yet run them on my PC. I'll run them once you've updated the README to include how to install the dependencies in a standard path i.e. `/usr/bin/` or `/usr/local/bin`.

**Comment by gnuhead-chieb** [comment]
>     1. Can we still build a pdf file?

No,To build pdf or something else, you have to convert to html once,then convert to else.



>     2\. Does every conversion process preserve the file structure?

Yes



>     3\. What's the state of the pdf files? Are the now converted into PNG or something?

These files are no meaning.I will remove them in finalizing process of this PR.

>     This should refer the list above, not the section.
I know about it.I'm looking for other implemention of this.

>     how the dependencies could be installed in both Linux and macOS. 
Most of these dependencies Just install by package manager.Then run required command on Terminal.It is OK if returned right output when run command on Terminal.

>     I don't know if xmllint is installed by default on Linux or not, macOS has it but it's way too outdated, you'd have to install it via Homebrew
xmllint is installed by default on most of Linux distros.This script doesn't using advanced feature of xmllint.so maybe works fine if it is outdated.

>     So, the only dependencies we should care about are pandoc-crossref,
If you want to install,
follow this guide
https://github.com/lierdakil/pandoc-crossref/blob/master/README.md

if not,
Download binary from [github releases](https://github.com/lierdakil/pandoc-crossref/releases) then put this on working directory(
AppManager/docs/raw/latex)

**Comment by MuntashirAkon** [comment]
I try very hard to make the project fully reproducible to anybody. So, kindly add the requirements in the build instruction section in the README first so that no confusion arises. I know how to install pandoc or any other dependencies but I will follow the README myself to make sure that it's understandable to others.

**Comment by gnuhead-chieb** [comment]
I added simple depschecker and install instruction（Debian based only,I dont use other distro or mac)

**Comment by MuntashirAkon** [comment]
I don't see any modification of the README file.

Notice that the conversions will be done using an automated script like it is now when you run `./gradlew assembleDebug`. If a required binary cannot be found, it would simply throw error since the script would return non-zero code. You don't have to check for binaries there (but it's good that you've done as it can be helpful). But you need to add the instructions in the build instructions section of the README file. This way, one would only need read the build instructions to know what to do/install at a first glance.

**Comment by gnuhead-chieb** [comment]
I added readme

**Comment by MuntashirAkon** [comment]
We already have png images at `/docs/raw/images` folder, those files in the tex files if pandoc supports it instead of uploading the new png images.

**Comment by MuntashirAkon** [comment]
Also, you are using GNU grep instead of the unix grep.

**Comment by gnuhead-chieb** [comment]
> We already have png images at `/docs/raw/images` folder, those files in the tex files if pandoc supports it instead of uploading the new png images.

Yes.However I made this script work only latex folder.
You can clone only latex folder using this command
`svn checkout https://github.com/gnuhead-chieb/AppManager/branches/gnuhead-chieb-patch-1/docs/raw/latex`



> Also, you are using GNU grep instead of the unix grep.

Fixed

**Comment by MuntashirAkon** [comment]
> Yes.However I made this script work only latex folder.

This should've very hard to configure.

Anyway, you've done a very good job, I think.

Here are my findings:

Screenshot 2021-05-16 at 00 15 09

We don't need bullet points, since we already have section numbering.

Screenshot 2021-05-16 at 00 09 39

Figure numbers do not match. In the figure, it says Figure 1 but in the text it says Figure 7.1.

Screenshot 2021-05-16 at 00 07 22

The links at the bottom should also be italicised. Use a common class name for the links if possible or make them bulleted.

Screenshot 2021-05-16 at 00 07 04

Screenshot 2021-05-16 at 00 06 48

Use of non-standard colours for errors and warnings.

Screenshot 2021-05-16 at 00 06 26

Remove the `>` sign. It looks odd and non-standard.

Apply the following changes to the `custom.css`:

```diff
diff --git a/docs/raw/latex/custom.css b/docs/raw/latex/custom.css
index a9446d74..396c8efb 100644
--- a/docs/raw/latex/custom.css
+++ b/docs/raw/latex/custom.css
@@ -35,7 +35,7 @@
     .main, .footnotes {max-width: 42em;}
 }
 
-.mdframed {
+.amalert {
     border-top: 0 !important;
     border-bottom: 0 !important;
     border-right: 0 !important;
@@ -44,6 +44,10 @@
     background: unset !important;
 }
 
+.amalert p:nth-child(1), .amalert p:nth-child(2) {
+    display: inline;
+}
+
 .itemize .colorbox {
     padding: 3px 7px;
     border-radius: 5px;
@@ -53,6 +57,10 @@ figcaption.caption span.id:after {
     content: ' ';
 }
 
+figure {
+    text-align: center;
+}
+
 @media (prefers-color-scheme: dark) {
     .sidenav {background: black !important;}
 
@@ -185,7 +193,6 @@ hr {
 }
 
 pre, code, kbd, samp {
-    color: #000;
     font-family: monospace, monospace;
     _font-family: 'courier new', monospace;
     font-size: 0.98em;

```

**Comment by gnuhead-chieb** [comment]
> We don't need bullet points, since we already have section numbering.

Added todo list.I will check it later



> Figure numbers do not match. In the figure, it says Figure 1 but in the text it says Figure 7.1.

Added todo list



> The links at the bottom should also be italicised. Use a common class name for the links if possible or make them bulleted.

Fixed!



> Use of non-standard colours for errors and warnings.

These color codes are taken from build output from make4ht.If you have better color,Please specify it.



> Remove the `>` sign. It looks odd and non-standard.

Fixed!



Did you checked translation process?Does it worked correctly?
And for currently,translation merging process is very slow.If you know tricks to faster this code,please let me know.

File: docs/raw/latex/appendices/app_ops.tex [hunk 1]
```diff
@@ -132,7 +132,7 @@ \subsection{PackageOps}\label{subsec:package-ops}
     \label{lst:package-ops-class}
 \end{listing}
 
-As can be seen in \Lref{lst:package-ops-class}, it stores all \textbf{OpEntry} for a package as well as the
+As can be seen in \Lref{subsec:package-ops}, it stores all \textbf{OpEntry} for a package as well as the
```
[the remark this answers is not in the document]
Comment (gnuhead-chieb) [reply docs/raw/latex/appendices/app_ops.tex:135]: Fixed!

**Comment by MuntashirAkon** [comment]
> These color codes are taken from build output from make4ht.If you have better color,Please specify it.

See here: https://muntashirakon.github.io/AppManager/en/#QQ2-1-42

> Did you checked translation process?Does it worked correctly?

They worked but for `faq_aot.tex` only.

> And for currently,translation merging process is very slow.If you know tricks to faster this code,please let me know.

I think most of the script can be rewritten using PHP. But don't worry about this right now.

**Comment by gnuhead-chieb** [comment]
> See here: https://muntashirakon.github.io/AppManager/en/#QQ2-1-42

Color updated

> They worked but for `faq_aot.tex` only.

This is why,I have added stringkeys for this file only for now.I will add stringkeys for all tex files.
Multifile support in doctool.sh is already covered(And this is also reason of made translation merging process very slow)
So I only have to do for translation is add stringkeys for all tex files

**Comment by MuntashirAkon** [comment]
> This is why,I have added stringkeys for this file only for now.I will add stringkeys for all tex files.
> Multifile support in doctool.sh is already covered(And this is also reason of made translation merging process very slow)
> So I only have to do for translation is add stringkeys for all tex files

I see. Please do add some comments on the script for clarity since it heavily uses Regex. I'll try to take care of the rests.

**Comment by MuntashirAkon** [comment]
Please resolve the conflicts, and briefly explain what to test here for me before finalising the docs.

**Comment by gnuhead-chieb** [comment]
I resolved conflicts.
Run commands wrote in [Readme](https://github.com/gnuhead-chieb/AppManager/blob/gnuhead-chieb-patch-1/docs/raw/latex/README)'s tutorial section and check if any bugs are appear.

**Comment by gnuhead-chieb** [comment]
added

**Review by MuntashirAkon — APPROVED** [review]

**Comment by MuntashirAkon** [comment]
@gnuhead-chieb: Can you write a script (simple bash script, for example) which converts a sample TEX file to XML and converts the XML back to TEX and show that both input and output files matched? The sample TEX file should consist of a few Latex commands as well as characters or symbols that needs conversion such as `<` (greater than), `>` (less than) and `&` (ampersand). Along with the sample TEX file, we will also test all the tex files located in the latex folder. I will include the script in tests so that we can be assured that new additions to the tex files aren't breaking anything.

**Comment by gnuhead-chieb** [comment]
> Can you write a script (simple bash script, for example) which converts a sample TEX file to XML and converts the XML back to TEX and show that both input and output files matched?

Remove `while read VARIABLE 〜 done < <(COMMAND)` on conversion process and replace variable to sample file's name.
Then just compare both files by command like `[[ $(cat INPUT) = $(cat OUTPUT) ]] && echo OK || echo Files are different`

**Comment by MuntashirAkon** [comment]
Well, after running `./doctool.sh updatetranslation` and `./doctool.sh mergetranslation`, the `.tex` files in the _latextranslated_ folder is expected to have the exact same contents as in the _latex_ folder. But it doesn't seem so. Some sections appear to be missing.

**Comment by gnuhead-chieb** [comment]
> Well, after running `./doctool.sh updatetranslation` and `./doctool.sh mergetranslation`, the `.tex` files in the _latextranslated_ folder is expected to have the exact same contents as in the _latex_ folder. But it doesn't seem so. Some sections appear to be missing.

fixed

**Comment by MuntashirAkon** [comment]
Some characters might not be translated correctly. For example, `&` is translated as `&amp;`.

**Comment by gnuhead-chieb** [comment]
fixed

**Comment by MuntashirAkon** [comment]
`appendices/app_ops.tex` still didn't match. At this point, use `sha*sum` (linux) or `shasum -a *` (macOS) to make sure that output tex matches with inputs. This should ensure that both the outputs are correctly being produced, and nothing is skipped. Did you check if Japanese texts are working correctly? Also, remove verbose logging from the `updatetranslation` (if not necessary) as printing makes operation a bit slow.

**Comment by gnuhead-chieb** [comment]
> appendices/app_ops.tex still didn't match. 

It seems the problem is what sed couldn't handle nested curly braces（{ }）
As written in readme(and I said before),this script extract title strings from `section{,subsection{,subsubsection{,chapter{,caption{,paragraph{` to next `}`.
appendices/app_ops.tex uses nested braces.
so it will cause bug.
ex)`\subsection{\texttt{OP\_*} Constants}`

But it has problem that it is difficult to handle nested string with regex.
If you know better way to handle nested braces and extract it, please let me know.

**Comment by gnuhead-chieb** [comment]
> Did you check if Japanese texts are working correctly?

Yes,but only for `faq/aot.tex`.I only translated this part right now.

> Also, remove verbose logging from the updatetranslation (if not necessary) as printing makes operation a bit slow.
Ok,I7ll remove it

**Comment by gnuhead-chieb** [comment]
For now,every file will be identical probably.
but nesting problem wasn't fixed yet.
Also,"cd $0" was disabled because it seems not working at all.

**Comment by MuntashirAkon** [comment]
> but nesting problem wasn't fixed yet.

What nesting problem?

Apart from other issues specified earlier, I've found a few minor issues:

1. There should be margin between two consecutive tips/note/warnings.
   Screenshot 2021-11-20 at 10 56 32
2. The see also section (not the inline ones) should contain bullet points
   Screenshot 2021-11-20 at 10 36 00
3. The stylesheet needs some changes in order to make the UI identical to the make4ht. We only need to edit custom.css, I believe.

**Comment by gnuhead-chieb** [comment]
> What nesting problem?

I wrote here
https://github.com/MuntashirAkon/AppManager/pull/418#issuecomment-974797406
and see string `appendices_appops-op-constants==title`

> The stylesheet needs some changes in order to make the UI identical to the make4ht. We only need to edit custom.css, I believe.
Yeah, but I'm not very familiar with CSS.I will investigate and fix them if I can. if not,please change them after this PR merged

**Comment by MuntashirAkon** [comment]
> I wrote here
> [#418 (comment)](https://github.com/MuntashirAkon/AppManager/pull/418#issuecomment-974797406)
> and see string `appendices_appops-op-constants==title`

I see. I think this is because regex cannot understand semantics, it can only be used for parsing. You'd need to some sort of memory elements (i.e. stacks) in order to keep track of the brackets, but that would make things a lot complicated (may be there's a antlr alternative in Python). I wonder why there's not a single project with proper support for localisation. At this point, I still maintain that localisation after converting latex to HTML is the best option available for us. I would've written a project myself if I had enough time after all the things that I do.

**Comment by gnuhead-chieb** [comment]
I will extract all title strings instead of extract strings in section tag where has nesting problem.(`%%#*string_key==titlex>>`)

ex) 
`\subsection{\texttt{OP\_*} Constants}\label{subsec:op-constants} %%#*appendices_appops-op-constants==titlex>>` 
will extracted as:
`<string name="appendices_changelogs-chapter==title"><![CDATA[\subsection{\texttt{OP\_*} Constants}\label{subsec:op-constants}]]></string>` 
in xliff.

> maintain that localisation after converting latex to HTML is the best option available for us

No,it more worse than this problem.Due to no string key,it will break all strings every resource update.Also,It hasn't more controls like option to choose not-translating strings.(see string intro_main-donation)

**Comment by gnuhead-chieb** [comment]
oh i found solution.nesting problem fixed

**Comment by MuntashirAkon** [comment]
> Due to no string key,it will break all strings every resource update.

Ah, I see.

**Comment by gnuhead-chieb** [comment]
> There should be margin between two consecutive tips/note/warnings.
fixed

**Comment by gnuhead-chieb** [comment]
done

**Comment by gnuhead-chieb** [comment]
@MuntashirAkon Please check last commit

**Comment by gnuhead-chieb** [comment]
is git rebase -i HEAD~n can squash commits?

**Pull request closed without merging**